[Term Entry] Neural Networks ai RMSprop (#4782)
Codecademy/docs: content/ai/concepts/neural-networks/terms/rmsprop/rmsprop.md

> ---
> Title: 'RMSprop'
> Description: 'RMSprop is an optimisation algorithm to adjust the learning rates dynamically for training neural networks.'
> Subjects:
>   - 'Machine Learning'
>   - 'Data Science'
>   - 'Computer Science'
> Tags:
>   - 'AI'
>   - 'Machine Learning'
>   - 'Neural Networks'
>   - 'Functions'
> CatalogContent:
>   - 'paths/machine-learning-ai-engineering-foundations'
>   - 'intro-to-ai-strategy'
> ---

Reviewer: ```suggestion
---
Title: 'RMSprop'
Description: 'RMSprop is an optimization algorithm that dynamically adjusts the learning rate during the training of neural networks.'
Subjects:
  - 'Computer Science'
  - 'Data Science'
  - 'Machine Learning'
Tags:
  - 'AI'
  - 'Functions'
  - 'Machine Learning'
  - 'Neural Networks'
CatalogContent:
  - 'paths/machine-learning-ai-engineering-foundations'
  - 'intro-to-ai-strategy'
---
```

Reviewer: Arranged the tags and subjects in an alphabetical manner. Reframed the description to be more readable.

**RMSprop** is an abbreviation for the term, **Root Mean Square Propagation**, and is an algorithm that is designed to adjust the learning rates dynamically for training neural network models. It helps to improve the stability and speed of the training process by adapting the learning rate based on recent gradients, making it effective for training deep neural networks.

## How Does RMSprop Work?

RMSprop modifies the learning rate for each parameter individually by taking into account the magnitudes of the recent gradients to adjust the updates. Here is how it works:

- **Exponential Decay of Squared Gradients**: RMSprop keeps track of the moving average of the squared gradients of the loss function with an exponential decay factor. The average is then used to normalise the gradients, preventing them from being too large or too small.
- **Update Rule**: For each parameter $\theta_i$, RMSprop updates the exponentially decaying average of the squared gradients $E[g^2]_t$ and then uses it to adjust the parameter:
  $$E[g^2]_t = \rho E[g^2]_{t-1} + (1-\rho)g^2_t$$

>   $$\theta_{t+1} = \theta_t - \frac{\eta}{\sqrt{E[g^2]_t + \epsilon}}g_t$$

Reviewer: Can this be represented either in an image or in a python equation format? Since this is a LaTeX equation, we need to present this in a different format.

  where $g_t$ refers to the gradient at time $t$, $\rho$ refers to the decay rate (which is typically around 0.9), $\eta$ refers to the learning rate, and $\epsilon$ refers to a small constant (around $10^{-8}$) to prevent any division by zero errors.

## Example

The following example demonstrates how RMSprop can be used for stochastic gradient descent:

```py

> import numpy as np
> 
> # Example neural network parameter
> theta = np.array([0.2, 0.4, 0.6])
> # Learning rate
> eta = 0.01
> # Decay rate
> rho = 0.9
> # Small constant to prevent zero division error
> epsilon = 1e-8
> # Running average of squared gradients
> E_g2 = np.zeros_like(theta)
> 
> # Example gradients from a loss function
> gradients = np.array([-0.5, 0.3, 0.24])
> 
> # Update running average
> E_g2 = rho * E_g2 + (1 - rho) * gradients**2
> 
> # Update parameters using RMSprop
> theta -= eta / np.sqrt(E_g2 + epsilon) * gradients
> 
> print("Updated parameters:", theta)

Reviewer: ```suggestion
import numpy as np

# Example neural network parameters (weights)
theta = np.array([0.2, 0.4, 0.6])

# Learning rate (step size for updating parameters)
eta = 0.01

# Decay rate (used to compute the moving average of squared gradients)
rho = 0.9

# Small constant to prevent division by zero
epsilon = 1e-8

# Running average of squared gradients
E_g2 = np.zeros_like(theta)

# Example gradients of the loss function with respect to the parameters
gradients = np.array([-0.5, 0.3, 0.24])

# Update running average of squared gradients
E_g2 = rho * E_g2 + (1 - rho) * gradients**2

# Update parameters using RMSprop
theta -= eta / np.sqrt(E_g2 + epsilon) * gradients

print("Updated parameters:", theta)
```

Reviewer: Reframed the comments for better understanding.

```

The output of the above code is as follows:

```shell
Updated parameters: [0.23162277 0.36837724 0.56837725]
```

## Codebyte Example

> The following Codebyte example also demonstrates the use of RMSprop in another stochastic gradient descent scenario:

Reviewer: ```suggestion
Run the following Codebyte to understand the use of RMSprop in a stochastic gradient descent scenario:
```
